### PR TITLE
PERF: StringEngine for string dtype indexing ops

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.3.0.rst` file if fixing a bug or adding a new feature.

Seems to be about 2x faster than `ObjectEngine` for indexing ops that use the hashmap:

```
import pandas as pd

N = 100_000
dtype = "string[pyarrow_numpy]"

strings = [f"i-{i}" for i in range(N)]

idx1 = pd.Index(strings[10:], dtype=dtype)
idx2 = pd.Index(strings[:-10], dtype=dtype)

%timeit idx1.get_indexer_for(idx2)

# 52.6 ms ± 834 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)  -> main
# 25 ms ± 790 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)    -> PR
```

### CODE REVIEW

File: pandas/_libs/index.pyi [hunk 1]
```diff
@@ -50,6 +50,7 @@ class UInt32Engine(IndexEngine): ...
 class UInt16Engine(IndexEngine): ...
 class UInt8Engine(IndexEngine): ...
 class ObjectEngine(IndexEngine): ...
+class StringEngine(ObjectEngine): ...
```
Comment (WillAyd) [inline pandas/_libs/index.pyi:53]: Think this needs to be updated to match the class implementation

**Review by WillAyd — COMMENTED** [review]
looks good to me

File: pandas/_libs/index.pyi (hunk 1 above)
Comment (lukemanley) [reply pandas/_libs/index.pyi]: Updated, thanks

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @lukemanley

**Pull request merged**